File: problem-of-the-day/day05/Design Twitter.py
class Twitter:
    def __init__(self):
        self.user_tweets = {}  # key = userId : value = deque of their tweets
        self.user_follows = {} # key = userId : value = set of whom they follow
        self.clock = 0 # even though self.user_tweets preserves order for a user, we need a clock to compare tweet times across users

    def postTweet(self, userId: int, tweetId: int) -> None:
        if userId not in self.user_tweets:
            self.user_tweets[userId] = deque()
        
        if userId not in self.user_follows:
            self.user_follows[userId] = set([userId])

        self.user_tweets[userId].append((self.clock, tweetId))
        self.clock += 1
        if len(self.user_tweets[userId]) > 10:
            self.user_tweets[userId].popleft() # discard the oldest tweet

    def getNewsFeed(self, userId: int) -> List[int]:
        if userId not in self.user_follows:
            self.user_follows[userId] = set([userId])

        n = len(self.user_follows[userId])
        index_to_pick = [-1] * n # tracks current most recent tweet
        feed = []

        # we want to pick 10 max elements from n sorted lists 
        while len(feed) < 10:
            max_time = -1
            max_tweet = None
            max_followee = None

            for i, followee in enumerate(self.user_follows[userId]):
                # if possible, consider the last index (most recent tweet) of each followee 
                if followee in self.user_tweets and (-index_to_pick[i]) <= len(self.user_tweets[followee]):
                    time, tweet = self.user_tweets[followee][index_to_pick[i]]
                    if time > max_time:
                        max_time = time
                        max_tweet = tweet
                        max_followee = i

            if max_followee is None:  # no more tweets
                break

            index_to_pick[max_followee] -= 1  # move to previous index in the list for this followee
            feed.append(max_tweet)

        return feed

    def follow(self, followerId: int, followeeId: int) -> None:
        if followerId not in self.user_follows:
            self.user_follows[followerId] = set([followerId])

        self.user_follows[followerId].add(followeeId)

    def unfollow(self, followerId: int, followeeId: int) -> None:
        if followerId in self.user_follows and followeeId in self.user_follows[followerId]:
            self.user_follows[followerId].remove(followeeId)

File: problem-of-the-day/day05/Design Twitter.cpp
class Twitter {
public:
    unordered_map<int,unordered_set<int>>ds;
    //This Data Structure will store list of twitter users and corresponding to each user there will be a set consiting of userid whom the user follow;
    vector<pair<int,int>>posts;
    //this vector will store the list of posts ::each post will have a tweetId and
    //the userId of the user Who posted it
    Twitter() {
        
    }
    
    void postTweet(int userId, int tweetId) {

        posts.push_back({tweetId,userId});
    }
    
    vector<int> getNewsFeed(int userId) {
        //If current user tries to retrieve 10 most recent post then he will be able to retrieve the posts that he himself posted or the post posted by another user whom current user follow
        vector<int>ans;

        //Iterate from last of posts since we want most recent
        int count=1;
        for(int i=posts.size()-1;i>=0&&count<=10;i--){
            int Post_UserId=posts[i].second;
            int post_Id=posts[i].first;
            if(Post_UserId==userId||ds[userId].find(Post_UserId)!=ds[userId].end()){
                ans.push_back(post_Id);
                count++;
            }
        }
        return ans;
    }
    
    void follow(int followerId, int followeeId) {
        //If A follows B then in a set corresponding to A  insert B
        ds[followerId].insert(followeeId);
    }
    
    void unfollow(int followerId, int followeeId) {
        //If A unfollows B then from a set corresponding to A remove B
        ds[followerId].erase(followeeId);
    }
};